stage0/src/main.rs

```diff
@@ -17,15 +17,139 @@
+        );
+    });
+}
+
+pub fn create_gdt(gdt: &mut GlobalDescriptorTable) -> (SegmentSelector, SegmentSelector) {
+    *gdt = GlobalDescriptorTable::new();
+    let cs = gdt.add_entry(Descriptor::kernel_code_segment());
+    let ds = gdt.add_entry(Descriptor::kernel_data_segment());
+    (cs, ds)
+}
+
+pub fn create_idt(idt: &mut InterruptDescriptorTable) {
+    *idt = InterruptDescriptorTable::new();
+}
+
+pub unsafe fn jump_to_kernel(address: VirtAddr) -> ! {
+    asm!(
+        // Boot stack pointer
+        "mov $0x8000, %rsp",
+        // Zero page address
+        "mov $0x7000, %rsi",
+        // ...and away we go!
+        "jmp *{0}",
+        in(reg) address.as_u64(),
+        options(noreturn, att_syntax)
+    );
+}
+
 #[no_mangle]
 pub extern "C" fn rust64_start() -> ! {
-    loop {
-        hlt();
+    /* Set up the machine according to the 64-bit Linux boot protocol.
+     * See https://www.kernel.org/doc/html/latest/x86/boot.html#id1 for the particular requirements.
+     */
+
+    // Safety, for all these data structure dereferences: The addresses are to well-known locations,
+    // same as crosvm is using, in the lowest 1MB of memory which we know is valid.
+    let gdt = unsafe { &mut *(BOOT_GDT_ADDR.as_u64() as *mut GlobalDescriptorTable) };
+    let (cs, ds) = create_gdt(gdt);
+    gdt.load();
+    // Safety: we've set up the valid data structures in create_gdt, above.
+    unsafe {
+        CS::set_reg(cs);
+        DS::set_reg(ds);
+        ES::set_reg(ds);
+        FS::set_reg(ds);
+        GS::set_reg(ds);
+        SS::set_reg(ds);
+    }
+
+    let idt = unsafe { &mut *(BOOT_IDT_ADDR.as_u64() as *mut InterruptDescriptorTable) };
+    create_idt(idt);
+    idt.load();
+
+    let pml4 = unsafe { &mut *(BOOT_PML4_ADDR.as_u64() as *mut PageTable) };
+    let pdpt = unsafe { &mut *(BOOT_PDPT_ADDR.as_u64() as *mut PageTable) };
+    let pd = unsafe { &mut *(BOOT_PD_ADDR.as_u64() as *mut PageTable) };
+    create_page_tables(pml4, pdpt, pd);
+    /* We need to do some trickery here. All of the stage0 code is somewhere within [4G-2M; 4G).
+     * Thus, let's keep our own last PD, so that we can continue executing after reloading the
+     * page tables.
+     */
+    // Safety: dereferencing the raw pointer is safe as that's the currently in-use page directory.
+    pdpt[3].set_addr(
+        PhysAddr::new((unsafe { &BIOS_PD as *const _ }) as u64),
+        PageTableFlags::PRESENT | PageTableFlags::WRITABLE,
+    );
+    // Safety: changing the page tables here is safe because (a) we've populated the new page tables
+    // at the well-known address of BOOT_PML4_ADDR and (b) we've ensured we keep the mapping for our
+    // own code.
+    unsafe {
+        Cr3::write(
+            PhysFrame::from_start_address(BOOT_PML4_ADDR).unwrap(),
+            Cr3Flags::empty(),
+        );
+    }
+    /* TODO: interrogate qemu for memory areas and set up the zero page. */
```

Comment:
I suspect some of the formatting code in CI will ask you to add issue numbers to the TODO statements.